src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/Meter.cs

```diff
@@ -227,6 +449,95 @@ public void Dispose()
             }
         }
 
+        private static bool CompareTags(IEnumerable<KeyValuePair<string, object?>>? tags1, IEnumerable<KeyValuePair<string, object?>>? tags2)
+        {
+            if (tags1 is null)
+            {
+                return tags2 is null;
+            }
+
+            if (tags2 is null)
+            {
+                return false;
+            }
+
+            if (tags1 is ICollection<KeyValuePair<string, object?>> firstCol && tags2 is ICollection<KeyValuePair<string, object?>> secondCol)
+            {
+                if (firstCol.Count != secondCol.Count)
+                {
+                    return false;
+                }
+
+                if (firstCol is IList<KeyValuePair<string, object?>> firstList && secondCol is IList<KeyValuePair<string, object?>> secondList)
+                {
+                    int count = firstList.Count;
+                    for (int i = 0; i < count; i++)
+                    {
+                        KeyValuePair<string, object?> pair1 = firstList[i];
+                        KeyValuePair<string, object?> pair2 = secondList[i];
+                        if (pair1.Key != pair2.Key || !object.Equals(pair1.Value, pair2.Value))
+                        {
+                            return false;
+                        }
+                    }
+
+                    return true;
+                }
+            }
+
+            using (IEnumerator<KeyValuePair<string, object?>> e1 = tags1.GetEnumerator())
+            using (IEnumerator<KeyValuePair<string, object?>> e2 = tags2.GetEnumerator())
+            {
+                while (e1.MoveNext())
+                {
+                    KeyValuePair<string, object?> pair1 = e1.Current;
+                    if (!e2.MoveNext() || pair1.Key != e2.Current.Key || !object.Equals(pair1.Value, e2.Current.Value))
+                    {
+                        return false;
+                    }
+                }
+
+                return !e2.MoveNext();
+            }
+        }
+
+        // AddInstrument will be called when publishing the instrument (i.e. calling Instrument.Publish()).
+        private Instrument GetOrCreateInstrument<T>(Type instrumentType, string name, string? unit, string? description, IEnumerable<KeyValuePair<string, object?>>? tags, Func<Instrument> instrumentCreator)
+        {
+            List<Instrument>? instrumentList;
+
+            lock (_nonObservableInstrumentsCache)
+            {
+                if (!_nonObservableInstrumentsCache.TryGetValue(name, out instrumentList))
+                {
+                    instrumentList = new List<Instrument>();
+                    _nonObservableInstrumentsCache.Add(name, instrumentList);
+                }
+            }
+
+            Debug.Assert(instrumentList != null);
+
+            lock (instrumentList)
+            {
+                foreach (Instrument instrument in instrumentList)
```

Comment:
In what situation is GetOrCreateInstrument called?  This is O(N) in the number of instruments; we'd want to avoid any situation where you might be doing this inside of some outer loop that would lead to quadratic behavior.

Reply:
This is O(N) of the instruments that have the same names. i.e. Not all instruments' lists. In the common cases we'll have only one instrument. I am not worried about that. 

> In what situation is GetOrCreateInstrument called?

This is called when creating a new instrument from this specific meter.